File: examples/api_example.php
<?php
/**
 * ============================================================
 * PROJOPP_AP_e3 - YouTube Downloader - Przykład API w PHP
 * Demonstracja integracji z backendem Node.js
 * 
 * Autor: konkowski2
 * Wersja: 1.00
 * Język: PHP
 * ============================================================
 * 
 * Ten plik pokazuje jak można zintegrować aplikację PHP
 * z naszym backendem Node.js do pobierania filmów YouTube.
 */

// ==================== KONFIGURACJA ====================

// URL backendu Node.js (domyślnie lokalny serwer)
// W produkcji zmień na właściwy adres serwera
define('API_URL', 'http://localhost:3001');

// Timeout dla żądań HTTP w sekundach
define('REQUEST_TIMEOUT', 30);

// Maksymalny czas oczekiwania na pobranie (5 minut)
define('MAX_WAIT_TIME', 300);

// ==================== KLASA OBSŁUGI API ====================

/**
 * Klasa YouTubeDownloader
 * 
 * Umożliwia komunikację z backendem Node.js
 * do pobierania filmów z YouTube
 * 
 * @package YouTubeDownloader
 * @version 1.00
 */
class YouTubeDownloader {
    
    // Przechowuje URL backendu
    private string $apiUrl;
    
    // Przechowuje ostatni błąd
    private ?string $lastError = null;
    
    /**
     * Konstruktor klasy
     * 
     * @param string $apiUrl - URL backendu (opcjonalny)
     */
    public function __construct(string $apiUrl = API_URL) {
        // Przypisz URL backendu do właściwości obiektu
        $this->apiUrl = rtrim($apiUrl, '/'); // Usuń trailing slash
    }
    
    /**
     * Rozpoczyna pobieranie filmu z YouTube
     * 
     * @param string $youtubeUrl - Link do filmu na YouTube
     * @param string $format - Format docelowy (mp4-best, mp4-1080, mp4-720, mp3, m4a)
     * @return string|null - ID zadania lub null w przypadku błędu
     * 
     * @example
     * $jobId = $downloader->startDownload('https://youtube.com/watch?v=xyz', 'mp3');
     */
    public function startDownload(string $youtubeUrl, string $format = 'mp4-best'): ?string {
        // Walidacja URL - sprawdź czy to YouTube
        if (!$this->isValidYoutubeUrl($youtubeUrl)) {
            $this->lastError = 'Nieprawidłowy link YouTube';
            return null;
        }
        
        // Walidacja formatu
        $allowedFormats = ['mp4-best', 'mp4-1080', 'mp4-720', 'mp3', 'm4a'];
        if (!in_array($format, $allowedFormats)) {
            $this->lastError = 'Nieobsługiwany format: ' . $format;
            return null;
        }
        
        // Przygotuj dane do wysłania (JSON)
        $postData = json_encode([
            'url' => $youtubeUrl,
            'format' => $format
        ]);
        
        // Wykonaj żądanie POST do backendu
        $response = $this->makeRequest('/api/start', 'POST', $postData);
        
        // Sprawdź czy żądanie się powiodło
        if ($response === null) {
            return null;
        }
        
        // Zdekoduj odpowiedź JSON
        $data = json_decode($response, true);
        
        // Zwróć ID zadania
        return $data['job_id'] ?? null;
    }
    
    /**
     * Sprawdza status zadania pobierania
     * 
     * @param string $jobId - ID zadania zwrócone przez startDownload()
     * @return array|null - Tablica ze statusem lub null przy błędzie
     * 
     * Status zawiera klucze:
     * - status: 'queued', 'downloading', 'processing', 'ready', 'error'
     * - progress: 0-100 (procent postępu)
     * - error: komunikat błędu (jeśli status == 'error')
     * - filename: nazwa pliku (jeśli status == 'ready')
     */
    public function checkStatus(string $jobId): ?array {
        // Walidacja ID zadania (powinno być 16-znakowym hexem)
        if (!preg_match('/^[a-f0-9]{16}$/', $jobId)) {
            $this->lastError = 'Nieprawidłowe ID zadania';
            return null;
        }
        
        // Wykonaj żądanie GET do backendu
        $response = $this->makeRequest('/api/status/' . $jobId, 'GET');
        
        if ($response === null) {
            return null;
        }
        
        // Zwróć zdekodowaną odpowiedź
        return json_decode($response, true);
    }
    
    /**
     * Czeka na zakończenie pobierania i zwraca URL do pliku
     * 
     * @param string $jobId - ID zadania
     * @param callable|null $progressCallback - Funkcja wywoływana przy zmianie postępu
     * @return string|null - URL do pobrania pliku lub null przy błędzie
     */
    public function waitForCompletion(string $jobId, ?callable $progressCallback = null): ?string {
        // Czas rozpoczęcia oczekiwania
        $startTime = time();
        
        // Pętla sprawdzająca status
        while (true) {
            // Sprawdź czy nie przekroczono timeout
            if (time() - $startTime > MAX_WAIT_TIME) {
                $this->lastError = 'Przekroczono maksymalny czas oczekiwania';
                return null;
            }
            
            // Pobierz aktualny status
            $status = $this->checkStatus($jobId);
            
            if ($status === null) {
                return null;
            }
            
            // Wywołaj callback z postępem (jeśli podany)
            if ($progressCallback !== null) {
                $progressCallback($status['status'], $status['progress'] ?? 0);
            }
            
            // Sprawdź czy zakończono
            switch ($status['status']) {
                case 'ready':
                    // Sukces - zwróć URL do pliku
                    return $this->apiUrl . '/api/file/' . $jobId;
                    
                case 'error':
                    // Błąd - zapisz komunikat
                    $this->lastError = $status['error'] ?? 'Nieznany błąd';
                    return null;
                    
                default:
                    // Jeszcze w trakcie - poczekaj sekundę
                    sleep(1);
            }
        }
    }
    
    /**
     * Waliduje URL YouTube
     * 
     * @param string $url - URL do sprawdzenia
     * @return bool - true jeśli URL jest prawidłowy
     */
    private function isValidYoutubeUrl(string $url): bool {
        // Wyrażenie regularne dla URL-i YouTube
        $pattern = '/^(https?:\/\/)?(www\.)?(youtube\.com|youtu\.be|m\.youtube\.com)\/.+/i';
        
        // Zwróć wynik dopasowania
        return (bool) preg_match($pattern, $url);
    }
    
    /**
     * Wykonuje żądanie HTTP do API
     * 
     * @param string $endpoint - Ścieżka endpointu (np. '/api/start')
     * @param string $method - Metoda HTTP ('GET' lub 'POST')
     * @param string|null $data - Dane do wysłania (dla POST)
     * @return string|null - Odpowiedź lub null przy błędzie
     */
    private function makeRequest(string $endpoint, string $method, ?string $data = null): ?string {
        // Pełny URL żądania
        $url = $this->apiUrl . $endpoint;
        
        // Konfiguracja kontekstu HTTP
        $options = [
            'http' => [
                'method' => $method,                    // Metoda HTTP
                'timeout' => REQUEST_TIMEOUT,           // Timeout
                'ignore_errors' => true,                // Nie rzucaj wyjątku przy błędach HTTP
                'header' => 'Content-Type: application/json' // Nagłówek JSON
            ]
        ];
        
        // Dodaj dane dla metody POST
        if ($method === 'POST' && $data !== null) {
            $options['http']['content'] = $data;
        }
        
        // Utwórz kontekst
        $context = stream_context_create($options);
        
        // Wykonaj żądanie
        $response = @file_get_contents($url, false, $context);
        
        // Sprawdź czy żądanie się powiodło
        if ($response === false) {
            $this->lastError = 'Nie można połączyć się z serwerem';
            return null;
        }
        
        return $response;
    }
    
    /**
     * Zwraca ostatni komunikat błędu
     * 
     * @return string|null - Komunikat błędu lub null
     */
    public function getLastError(): ?string {
        return $this->lastError;
    }
}

// ==================== PRZYKŁAD UŻYCIA ====================

// Sprawdź czy skrypt uruchomiony z CLI
if (php_sapi_name() === 'cli') {
    
    echo "=== YouTube Downloader - PHP Client ===\n\n";
    
    // Utwórz instancję downloadera
    $downloader = new YouTubeDownloader();
    
    // Przykładowy URL (zmień na prawdziwy)
    $testUrl = 'https://www.youtube.com/watch?v=dQw4w9WgXcQ';
    
    echo "Rozpoczynam pobieranie: $testUrl\n";
    echo "Format: MP3\n\n";
    
    // Rozpocznij pobieranie
    $jobId = $downloader->startDownload($testUrl, 'mp3');
    
    if ($jobId === null) {
        // Błąd - wyświetl komunikat
        echo "BŁĄD: " . $downloader->getLastError() . "\n";
        exit(1);
    }
    
    echo "ID zadania: $jobId\n";
    echo "Oczekiwanie na zakończenie...\n\n";
    
    // Callback do wyświetlania postępu
    $progressCallback = function(string $status, int $progress) {
        // Wyświetl pasek postępu
        $bar = str_repeat('█', (int)($progress / 5));
        $empty = str_repeat('░', 20 - strlen($bar));
        echo "\r[$bar$empty] $progress% - $status";
    };
    
    // Czekaj na zakończenie
    $downloadUrl = $downloader->waitForCompletion($jobId, $progressCallback);
    
    echo "\n\n";
    
    if ($downloadUrl !== null) {
        echo "SUKCES! Plik dostępny pod adresem:\n";
        echo $downloadUrl . "\n";
    } else {
        echo "BŁĄD: " . $downloader->getLastError() . "\n";
        exit(1);
    }
}

?>

File: examples/api_example.py
"""
============================================================
PROJOPP_AP_e3 - YouTube Downloader - Przykład API w Python
Demonstracja integracji z backendem Node.js

Autor: konkowski2
Wersja: 1.00
Język: Python 3
============================================================

Ten plik pokazuje jak można zintegrować aplikację Python
z naszym backendem Node.js do pobierania filmów YouTube.
"""

# ==================== IMPORTY ====================

# Moduł requests do wykonywania żądań HTTP
# Alternatywnie można użyć wbudowanego urllib
import requests

# Moduł time do opóźnień (sleep)
import time

# Moduł re do wyrażeń regularnych (walidacja URL)
import re

# Moduł sys do argumentów wiersza poleceń
import sys

# Typing dla typowania (Python 3.5+)
from typing import Optional, Dict, Callable, Any

# ==================== KONFIGURACJA ====================

# URL backendu Node.js (domyślnie lokalny serwer)
API_URL = "http://localhost:3001"

# Timeout dla żądań HTTP w sekundach
REQUEST_TIMEOUT = 30

# Maksymalny czas oczekiwania na pobranie (5 minut)
MAX_WAIT_TIME = 300

# Dozwolone formaty pobierania
ALLOWED_FORMATS = ["mp4-best", "mp4-1080", "mp4-720", "mp3", "m4a"]


# ==================== KLASA DOWNLOADERA ====================

class YouTubeDownloader:
    """
    Klasa do komunikacji z backendem YouTube Downloader.
    
    Umożliwia:
    - Rozpoczęcie pobierania filmu z YouTube
    - Sprawdzanie statusu pobierania
    - Oczekiwanie na zakończenie z callback'iem postępu
    
    Attributes:
        api_url (str): URL backendu API
        last_error (str): Ostatni komunikat błędu
        
    Example:
        >>> downloader = YouTubeDownloader()
        >>> job_id = downloader.start_download("https://youtube.com/watch?v=xyz", "mp3")
        >>> url = downloader.wait_for_completion(job_id)
    """
    
    def __init__(self, api_url: str = API_URL) -> None:
        """
        Inicjalizacja downloadera.
        
        Args:
            api_url: URL backendu API (domyślnie localhost:3001)
        """
        # Przypisz URL backendu (usuń trailing slash)
        self.api_url = api_url.rstrip("/")
        
        # Inicjalizuj zmienną na ostatni błąd
        self.last_error: Optional[str] = None
        
        # Utwórz sesję requests (reużywa połączenia HTTP)
        self._session = requests.Session()
    
    def start_download(self, youtube_url: str, format: str = "mp4-best") -> Optional[str]:
        """
        Rozpoczyna pobieranie filmu z YouTube.
        
        Args:
            youtube_url: Link do filmu na YouTube
            format: Format docelowy (mp4-best, mp4-1080, mp4-720, mp3, m4a)
            
        Returns:
            ID zadania (string) lub None w przypadku błędu
            
        Raises:
            Nie rzuca wyjątków - błędy zapisywane w last_error
        """
        # Walidacja URL - sprawdź czy to prawidłowy link YouTube
        if not self._is_valid_youtube_url(youtube_url):
            self.last_error = "Nieprawidłowy link YouTube"
            return None
        
        # Walidacja formatu - sprawdź czy jest na liście dozwolonych
        if format not in ALLOWED_FORMATS:
            self.last_error = f"Nieobsługiwany format: {format}"
            return None
        
        # Przygotuj dane do wysłania
        payload = {
            "url": youtube_url,
            "format": format
        }
        
        try:
            # Wykonaj żądanie POST do backendu
            response = self._session.post(
                f"{self.api_url}/api/start",
                json=payload,                # Automatycznie serializuje do JSON
                timeout=REQUEST_TIMEOUT
            )
            
            # Sprawdź czy odpowiedź jest poprawna (kod 2xx)
            if not response.ok:
                # Próbuj odczytać błąd z odpowiedzi
                error_data = response.json()
                self.last_error = error_data.get("error", f"Błąd HTTP {response.status_code}")
                return None
            
            # Odczytaj i zwróć ID zadania
            data = response.json()
            return data.get("job_id")
            
        except requests.RequestException as e:
            # Błąd połączenia (timeout, brak sieci, etc.)
            self.last_error = f"Błąd połączenia: {str(e)}"
            return None
    
    def check_status(self, job_id: str) -> Optional[Dict[str, Any]]:
        """
        Sprawdza status zadania pobierania.
        
        Args:
            job_id: ID zadania zwrócone przez start_download()
            
        Returns:
            Słownik ze statusem lub None przy błędzie
            
        Klucze w zwracanym słowniku:
            - status: 'queued', 'downloading', 'processing', 'ready', 'error'
            - progress: 0-100 (procent postępu)
            - error: komunikat błędu (jeśli status == 'error')
            - filename: nazwa pliku (jeśli status == 'ready')
        """
        # Walidacja ID zadania (powinno być 16-znakowym hexem)
        if not re.match(r'^[a-f0-9]{16}$', job_id):
            self.last_error = "Nieprawidłowe ID zadania"
            return None
        
        try:
            # Wykonaj żądanie GET do backendu
            response = self._session.get(
                f"{self.api_url}/api/status/{job_id}",
                timeout=REQUEST_TIMEOUT
            )
            
            if not response.ok:
                error_data = response.json()
                self.last_error = error_data.get("error", f"Błąd HTTP {response.status_code}")
                return None
            
            # Zwróć zdekodowaną odpowiedź
            return response.json()
            
        except requests.RequestException as e:
            self.last_error = f"Błąd połączenia: {str(e)}"
            return None
    
    def wait_for_completion(
        self, 
        job_id: str, 
        progress_callback: Optional[Callable[[str, int], None]] = None
    ) -> Optional[str]:
        """
        Czeka na zakończenie pobierania i zwraca URL do pliku.
        
        Args:
            job_id: ID zadania
            progress_callback: Funkcja wywoływana przy zmianie postępu
                               Przyjmuje (status: str, progress: int)
                               
        Returns:
            URL do pobrania pliku lub None przy błędzie
        """
        # Czas rozpoczęcia oczekiwania
        start_time = time.time()
        
        # Pętla sprawdzająca status
        while True:
            # Sprawdź czy nie przekroczono timeout
            elapsed = time.time() - start_time
            if elapsed > MAX_WAIT_TIME:
                self.last_error = "Przekroczono maksymalny czas oczekiwania"
                return None
            
            # Pobierz aktualny status
            status = self.check_status(job_id)
            
            if status is None:
                return None  # Błąd już zapisany w last_error
            
            # Wywołaj callback z postępem (jeśli podany)
            if progress_callback is not None:
                progress_callback(status["status"], status.get("progress", 0))
            
            # Sprawdź status zadania
            current_status = status["status"]
            
            if current_status == "ready":
                # Sukces - zwróć URL do pliku
                return f"{self.api_url}/api/file/{job_id}"
            
            elif current_status == "error":
                # Błąd - zapisz komunikat
                self.last_error = status.get("error", "Nieznany błąd")
                return None
            
            else:
                # Jeszcze w trakcie - poczekaj sekundę
                time.sleep(1)
    
    def _is_valid_youtube_url(self, url: str) -> bool:
        """
        Waliduje URL YouTube.
        
        Args:
            url: URL do sprawdzenia
            
        Returns:
            True jeśli URL jest prawidłowym linkiem YouTube
        """
        # Wyrażenie regularne dla URL-i YouTube
        # Akceptuje: youtube.com, youtu.be, m.youtube.com
        pattern = r'^(https?://)?(www\.)?(youtube\.com|youtu\.be|m\.youtube\.com)/.+'
        
        # Zwróć wynik dopasowania (case-insensitive)
        return bool(re.match(pattern, url, re.IGNORECASE))
    
    def get_last_error(self) -> Optional[str]:
        """
        Zwraca ostatni komunikat błędu.
        
        Returns:
            Komunikat błędu lub None jeśli nie było błędu
        """
        return self.last_error


# ==================== FUNKCJE POMOCNICZE ====================

def print_progress_bar(status: str, progress: int) -> None:
    """
    Wyświetla pasek postępu w terminalu.
    
    Args:
        status: Aktualny status ('downloading', 'processing', etc.)
        progress: Postęp w procentach (0-100)
    """
    # Oblicz długość wypełnionej części paska
    filled_length = int(20 * progress / 100)
    
    # Utwórz pasek
    bar = '█' * filled_length + '░' * (20 - filled_length)
    
    # Wyświetl pasek (nadpisz poprzednią linię)
    print(f'\r[{bar}] {progress:3d}% - {status}', end='', flush=True)


def main() -> None:
    """
    Główna funkcja programu - demonstracja użycia YouTubeDownloader.
    """
    print("=== YouTube Downloader - Python Client ===\n")
    
    # Utwórz instancję downloadera
    downloader = YouTubeDownloader()
    
    # Pobierz URL z argumentów lub użyj domyślnego
    if len(sys.argv) > 1:
        test_url = sys.argv[1]
    else:
        # Przykładowy URL (zmień na prawdziwy)
        test_url = "https://www.youtube.com/watch?v=dQw4w9WgXcQ"
    
    # Pobierz format z argumentów lub użyj domyślnego
    format_type = sys.argv[2] if len(sys.argv) > 2 else "mp3"
    
    print(f"URL: {test_url}")
    print(f"Format: {format_type}\n")
    
    # Rozpocznij pobieranie
    print("Rozpoczynam pobieranie...")
    job_id = downloader.start_download(test_url, format_type)
    
    if job_id is None:
        # Błąd - wyświetl komunikat i zakończ
        print(f"\nBŁĄD: {downloader.get_last_error()}")
        sys.exit(1)
    
    print(f"ID zadania: {job_id}")
    print("Oczekiwanie na zakończenie...\n")
    
    # Czekaj na zakończenie z paskiem postępu
    download_url = downloader.wait_for_completion(job_id, print_progress_bar)
    
    print("\n")  # Nowa linia po pasku postępu
    
    if download_url is not None:
        print("SUKCES! Plik dostępny pod adresem:")
        print(download_url)
    else:
        print(f"BŁĄD: {downloader.get_last_error()}")
        sys.exit(1)


# ==================== PUNKT WEJŚCIA ====================

# Sprawdź czy skrypt uruchomiony bezpośrednio (nie importowany)
if __name__ == "__main__":
    main()
